pkg/controller/validatingadmissionpolicystatus/crd.go

```diff
@@ -0,0 +1,249 @@
+		dependencies = set.UnsortedList()
+	} else {
+		return
+	}
+	delete(t.policyDependencies, policyName)
+	for _, gvk := range dependencies {
+		if set, ok := t.affectedPolicies[gvk]; ok {
+			set.Delete(policyName)
+			if set.Len() == 0 {
+				delete(t.affectedPolicies, gvk)
+			}
+		}
+	}
+}
+
+func CRDVersionToGVK(crd *apiextensionsv1.CustomResourceDefinition, crdVersion *apiextensionsv1.CustomResourceDefinitionVersion) schema.GroupVersionKind {
+	gvk := schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: crdVersion.Name,
+		Kind:    crd.Spec.Names.Kind,
+	}
+	return gvk
+}
+
+type cachedTypeResolver struct {
+	meta.RESTMapper
+	resolver.SchemaResolver
+	crdClient apiextensionsv1client.CustomResourceDefinitionInterface
+
+	cachedCRD    map[schema.GroupResource]*apiextensionsv1.CustomResourceDefinition
+	cachedSchema map[schema.GroupVersionKind]*spec.Schema
+}
+
+func newCachedTypeResolver(restMapper meta.RESTMapper, resolver resolver.SchemaResolver, crdClient apiextensionsv1client.CustomResourceDefinitionInterface) *cachedTypeResolver {
+	return &cachedTypeResolver{
+		RESTMapper:     restMapper,
+		SchemaResolver: resolver,
+		crdClient:      crdClient,
+		cachedCRD:      map[schema.GroupResource]*apiextensionsv1.CustomResourceDefinition{},
+		cachedSchema:   map[schema.GroupVersionKind]*spec.Schema{},
+	}
+}
+
+func (r *cachedTypeResolver) ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error) {
+	if s, ok := r.cachedSchema[gvk]; ok {
+		return s, nil
+	}
+	return r.SchemaResolver.ResolveSchema(gvk)
+}
+
+func (r *cachedTypeResolver) KindsFor(resource schema.GroupVersionResource) ([]schema.GroupVersionKind, error) {
+	if isGroupBuiltin(resource.Group) {
+		return r.RESTMapper.KindsFor(resource)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	crd, ok := r.cachedCRD[resource.GroupResource()]
+	if !ok {
+		var err error
+		// Assuming resource.Resource is plural.
+		// There does not seem to be a good way to reliably to convert the name without the CRD itself
+		crdName := resource.Resource + "." + resource.Group
+		crd, err = r.crdClient.Get(ctx, crdName, metav1.GetOptions{})
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				// also keep negative cache
+				r.cachedCRD[resource.GroupResource()] = nil
+				return nil, nil
+			}
+			return nil, err
+		}
+		r.cachedCRD[resource.GroupResource()] = crd
+
+		for _, v := range crd.Spec.Versions {
+			gvk := CRDVersionToGVK(crd, &v)
+			s, err := convertCRDValidationToSchema(v.Schema)
+			if err != nil {
+				return nil, err
+			}
+			r.cachedSchema[gvk] = s
+		}
+	}
+
+	var gvks []schema.GroupVersionKind
+	for _, v := range crd.Spec.Versions {
+		gvk := CRDVersionToGVK(crd, &v)
+		if v.Name == gvk.Version {
+			gvks = append(gvks, gvk)
+		}
+	}
+	return gvks, nil
+}
+
+func (r *cachedTypeResolver) ReferredGroupResources() []schema.GroupResource {
+	var grs []schema.GroupResource
```

Comment:
tiny nit: pre-allocate slice to approperiate size since it is known?

Reply:
Good catch! Fixed.